design_docs/mpas_xarray_class.md

```diff
@@ -0,0 +1,185 @@
+
+<h2> Requirement: MPAS-Analysis specific functionality should be supported in
+xarray preprossing <br>
+Date last modified: 2017/02/10 <br>
+Contributors: Xylar Asay-Davis
+</h2>
+
+There should be a way to perform MPAS-Analysis specific functionality such as
+mapping variables during preprocessing.  This functionality should be
+relatively easy to add to as new preprocessing needs arise.
+
+
+<h1> Algorithmic Formulations (optional) </h1>
+
+<h2> Algorithm: mpas_xarray does not include MPAS-Analysis specific
+functionality <br>
+Date last modified: 2017/02/10 <br>
+Contributors: Xylar Asay-Davis
+</h2>
+
+All functions and function arguments related to variable mapping will
+be removed from mpas_xarray and moved elsewhere.
+
+<h2> Algorithm: MPAS-Analysis specific functionality should be supported in
+xarray preprossing <br>
+Date last modified: 2017/02/10 <br>
+Contributors: Xylar Asay-Davis
+</h2>
+
+While this requirement seems almost contradictory to the previous
+requirement, both can be accommodated by using classes and inheritance.
+An MpasXarray class will be added to mpas_xarray, including a method
+(`_preprocess()`) for performing preprocessing.  Then, a subclass will
+be created (MpasAnalysisXarray) that includes the functionality that
+is specific to MPAS-Analysis by overloading the preprocessing method
+from MpasXarray.
+
+This solution may add a level of complexity, particularly for
+developing mpas_xarray and its analysis-specific derived reader.
+It also risks confusing a new developer by having seemingly redundant
+functionality in two places.  However, it seems to me to be the
+cleanest way to satisfy both requirements.
+
+
+<h1> Design and Implementation </h1>
+
+<h2> Implementation: mpas_xarray does not include MPAS-Analysis specific
+functionality <br>
+Date last modified: 2017/02/10 <br>
+Contributors: Xylar Asay-Davis
+</h2>
+
+A test branch can be found here 
+[xylar/MPAS-Analysis/variable_mapping_reorg](https://github.com/xylar/MPAS-Analysis/tree/variable_mapping_reorg)
+
+I have removed `map_variable` and `rename_variables` from `mpas_xarray`.
+I also removed any mention of the variable map from the rest of `mpas_xarray`.
+
+This branch also includes several other cleanup operations that are not
+addressing any requirements.  These include:
+   - I added a new helper function, `open_multifile_dataset`, for opening an
+     xarray data set in a single, simple command without reference to the
+     preprocessor.  This function should make opening new data sets more
+     intuitive for mpas_xarray users.  Calling the 
+     `MpasXarray.open_multifile_dataset` allows a bit more flexibility at the
+     cost of being a little bit less compact.
+   - making several utility functions non-public (it is unclear to me why anyone
+     want to call these directly):
+       - `_assert_valid_datetimes`
+       - `_assert_valid_selections`
+       - `_ensure_list`
+       - `_get_datetimes`
+       - `_remove_repeated_time_index`
```

Comment:
We may actually want this public, e.g., if we load and join two separate parts of a dataset at the user-level.